File: Algorithms/Implementation/Modified Kaprekar Numbers/java/Kaprekar.java
import java.util.Arrays;
import java.util.Scanner;
import java.lang.Math;

class Kaprekar {
	private static int digits(Long number) {
		int count = 0;
		while (number != 0) {
			number /= 10;
			count++;
		}
		return count;
	}

	private static void kaprekarNumbers(int p, int q) {
		int k_capacity = 30;
		int k_size = 0;
		int[] kaprekar = new int[k_capacity];

		for (int i = p; i <= q; i++) {
			int right_size = digits(Long.valueOf(i));
			Long square = (long)Math.pow(i, 2);
			int left_size = digits(square) - right_size;

			int left = (int)(square / Math.pow(10, right_size));
			int right = (int)(square % Math.pow(10, right_size));
			int sum = left + right;

			if (sum == i) {
				if (k_size+1 > k_capacity) {
					k_capacity *= 2;
					Arrays.copyOf(kaprekar, k_capacity);
				}
				kaprekar[k_size++] = i;
			}
		}

		if (k_size == 0) {
			System.out.println("INVALID RANGE");
		} else {
			for (int i = 0; i < k_size; i++) {
				System.out.print(i == k_size-1 ? kaprekar[i] : kaprekar[i] + " ");
			}
		}
	}

	public static void main(String[] args) {
		Scanner input = new Scanner(System.in);
		int p = input.nextInt();
		int q = input.nextInt();
		kaprekarNumbers(p, q);
	}
}

File: Algorithms/Implementation/Modified Kaprekar Numbers/python/kaprekar.py
class Kaprekar():
    def __init__(self, p, q):
        self.__kaprekar = list()
        self.__KaprekarNumbers(p, q)

    def __Digits(self, number):
        count = 0

        while number != 0:
            number = int(number / 10)
            count += 1
            
        return count

    def __KaprekarNumbers(self, p, q):
        for i in range(p, q+1):
            right_size = self.__Digits(i)
            square = int(pow(i, 2))
            left_size = self.__Digits(square) - right_size
            square = str(square)
#            print('i=' + str(i) + ", square=" + square + ", left_size=" + str(left_size) + ", right_size=" + str(right_size))
            left = 0

            if left_size > 0:
                left = int(square[:left_size])

            right = int(square[left_size:])
            sum = left + right
#            print('left=' + str(left) + ", right=" + str(right) + ", sum=" + str(sum))

            if sum == i:
                self.__kaprekar.append(i)

    def Show(self):
        if len(self.__kaprekar) == 0:
            print("INVALID RANGE")
        else:
            for i in range(len(self.__kaprekar)):
                print(str(self.__kaprekar[i])
                    if i == len(self.__kaprekar)-1
                    else str(self.__kaprekar[i]) + " ", end='')